Cannot build package on Centos 7

Hi!
> sessionInfo()
R version 3.5.0 (2018-04-23)
Platform: x86_64-redhat-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)


Try install websocket package and got error:

> devtools::install_github("rstudio/websocket")
Downloading GitHub repo rstudio/websocket@master
from URL https://api.github.com/repos/rstudio/websocket/zipball/master
Installing websocket
'/usr/lib64/R/bin/R' --no-site-file --no-environ --no-save --no-restore  \
  --quiet CMD INSTALL  \
  '/tmp/Rtmp7RR1Xj/devtools710c30061c10/rstudio-websocket-7b9761c'  \
  --library='/usr/lib64/R/library' --install-tests

* installing *source* package ‘websocket’ ...
Found pkg-config cflags and libs!
Using PKG_CFLAGS=
Using PKG_LIBS=-l:libssl.so.1.0.2k -l:libcrypto.so.1.0.2k
** libs
g++ -m64 -std=gnu++11 -I"/usr/include/R" -DNDEBUG -I./lib   -I"/usr/lib64/R/library/Rcpp/include" -I"/usr/lib64/R/library/BH/include" -I"/usr/lib64/R/library/AsioHeaders/include" -I/usr/local/include   -fpic  -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -c RcppExports.cpp -o RcppExports.o
g++ -m64 -std=gnu++11 -I"/usr/include/R" -DNDEBUG -I./lib   -I"/usr/lib64/R/library/Rcpp/include" -I"/usr/lib64/R/library/BH/include" -I"/usr/lib64/R/library/AsioHeaders/include" -I/usr/local/include   -fpic  -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -c websocket.cpp -o websocket.o
websocket.cpp: In function ‘SEXPREC* wsCreate(std::string, Rcpp::Environment, Rcpp::Environment, Rcpp::CharacterVector, Rcpp::CharacterVector)’:
websocket.cpp:212:59: error: invalid conversion from ‘int’ to ‘Rboolean’ [-fpermissive]
   R_RegisterCFinalizerEx(client_xptr, client_deleter, TRUE);
                                                           ^
In file included from /usr/lib64/R/library/Rcpp/include/Rcpp/r/headers.h:53:0,
                 from /usr/lib64/R/library/Rcpp/include/RcppCommon.h:29,
                 from /usr/lib64/R/library/Rcpp/include/Rcpp.h:27,
                 from websocket.cpp:1:
/usr/include/R/Rinternals.h:1093:6: error:   initializing argument 3 of ‘void R_RegisterCFinalizerEx(SEXP, R_CFinalizer_t, Rboolean)’ [-fpermissive]
 void R_RegisterCFinalizerEx(SEXP s, R_CFinalizer_t fun, Rboolean onexit);
      ^
websocket.cpp: In function ‘void wsSend(SEXP, SEXP)’:
websocket.cpp:261:72: error: invalid conversion from ‘int’ to ‘Rboolean’ [-fpermissive]
     int len = R_nchar(STRING_ELT(msg, 0), Bytes, FALSE, FALSE, "wsSend");
                                                                        ^
In file included from /usr/lib64/R/library/Rcpp/include/Rcpp/r/headers.h:53:0,
                 from /usr/lib64/R/library/Rcpp/include/RcppCommon.h:29,
                 from /usr/lib64/R/library/Rcpp/include/Rcpp.h:27,
                 from websocket.cpp:1:
/usr/include/R/Rinternals.h:999:5: error:   initializing argument 3 of ‘int R_nchar(SEXP, nchar_type, Rboolean, Rboolean, const char*)’ [-fpermissive]
 int R_nchar(SEXP string, nchar_type type_,
     ^
websocket.cpp:261:72: error: invalid conversion from ‘int’ to ‘Rboolean’ [-fpermissive]
     int len = R_nchar(STRING_ELT(msg, 0), Bytes, FALSE, FALSE, "wsSend");
                                                                        ^
In file included from /usr/lib64/R/library/Rcpp/include/Rcpp/r/headers.h:53:0,
                 from /usr/lib64/R/library/Rcpp/include/RcppCommon.h:29,
                 from /usr/lib64/R/library/Rcpp/include/Rcpp.h:27,
                 from websocket.cpp:1:
/usr/include/R/Rinternals.h:999:5: error:   initializing argument 4 of ‘int R_nchar(SEXP, nchar_type, Rboolean, Rboolean, const char*)’ [-fpermissive]
 int R_nchar(SEXP string, nchar_type type_,
     ^
In file included from /usr/lib64/R/library/AsioHeaders/include/asio/detail/impl/posix_mutex.ipp:24:0,
                 from /usr/lib64/R/library/AsioHeaders/include/asio/detail/posix_mutex.hpp:71,
                 from /usr/lib64/R/library/AsioHeaders/include/asio/detail/mutex.hpp:25,
                 from /usr/lib64/R/library/AsioHeaders/include/asio/detail/service_registry.hpp:20,
                 from /usr/lib64/R/library/AsioHeaders/include/asio/impl/execution_context.hpp:20,
                 from /usr/lib64/R/library/AsioHeaders/include/asio/execution_context.hpp:406,
                 from /usr/lib64/R/library/AsioHeaders/include/asio/detail/scheduler.hpp:21,
                 from /usr/lib64/R/library/AsioHeaders/include/asio/system_executor.hpp:19,
                 from /usr/lib64/R/library/AsioHeaders/include/asio/associated_executor.hpp:21,
                 from /usr/lib64/R/library/AsioHeaders/include/asio.hpp:19,
                 from ./lib/websocketpp/common/asio.hpp:52,
                 from ./lib/websocketpp/transport/asio/base.hpp:31,
                 from ./lib/websocketpp/transport/asio/connection.hpp:31,
                 from ./lib/websocketpp/transport/asio/endpoint.hpp:32,
                 from ./lib/websocketpp/config/asio_no_tls_client.hpp:32,
                 from client.hpp:4,
                 from websocket.cpp:24:
/usr/lib64/R/library/AsioHeaders/include/asio/error.hpp: At global scope:
/usr/lib64/R/library/AsioHeaders/include/asio/error.hpp:256:36: warning: ‘asio::error::system_category’ defined but not used [-Wunused-variable]
 static const asio::error_category& system_category
                                    ^
/usr/lib64/R/library/AsioHeaders/include/asio/error.hpp:258:36: warning: ‘asio::error::netdb_category’ defined but not used [-Wunused-variable]
 static const asio::error_category& netdb_category
                                    ^
/usr/lib64/R/library/AsioHeaders/include/asio/error.hpp:260:36: warning: ‘asio::error::addrinfo_category’ defined but not used [-Wunused-variable]
 static const asio::error_category& addrinfo_category
                                    ^
/usr/lib64/R/library/AsioHeaders/include/asio/error.hpp:262:36: warning: ‘asio::error::misc_category’ defined but not used [-Wunused-variable]
 static const asio::error_category& misc_category
                                    ^
In file included from /usr/lib64/R/library/AsioHeaders/include/asio/ssl/impl/context.ipp:25:0,
                 from /usr/lib64/R/library/AsioHeaders/include/asio/ssl/context.hpp:755,
                 from /usr/lib64/R/library/AsioHeaders/include/asio/ssl.hpp:18,
                 from ./lib/websocketpp/common/asio_ssl.hpp:34,
                 from ./lib/websocketpp/transport/asio/security/tls.hpp:35,
                 from ./lib/websocketpp/config/asio_client.hpp:33,
                 from client.hpp:5,
                 from websocket.cpp:24:
/usr/lib64/R/library/AsioHeaders/include/asio/ssl/error.hpp:33:36: warning: ‘asio::error::ssl_category’ defined but not used [-Wunused-variable]
 static const asio::error_category& ssl_category
                                    ^
make: *** [websocket.o] Error 1
ERROR: compilation failed for package ‘websocket’
* removing ‘/usr/lib64/R/library/websocket’
Installation failed: Command failed (1)


>gcc --version
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-28)


also I try with gcc 7.1, and got compile error again

--- a/src/client.hpp
+++ b/src/client.hpp
@@ -5,6 +5,12 @@
 #include <websocketpp/config/asio_client.hpp>
 #include <websocketpp/client.hpp>
 
+// This is necessary on CentOS 7, where bringing in asio_client.hpp indirectly loads
+// krb5/krb5.h, which defines TRUE/FALSE as integers. This then breaks R, as it
+// relies on TRUE/FALSE being part of the Rboolean enum.
+#undef TRUE
+#undef FALSE
+
 
 typedef ws_websocketpp::config::asio_client::message_type::ptr message_ptr;
 typedef ws_websocketpp::lib::function<void(ws_websocketpp::connection_hdl, message_ptr)> message_handler;

